internal/scheduler/scheduler.go

```diff
@@ -0,0 +1,264 @@
+	}
+	if logger == nil {
+		return nil, errors.New(errors.InvalidParameter, op, "missing logger")
+	}
+
+	opts := getOpts(opt...)
+	return &Scheduler{
+		serverId:        serverId,
+		jobRepoFn:       jobRepoFn,
+		logger:          logger,
+		registeredJobs:  new(sync.Map),
+		runningJobs:     make(map[JobId]runningJob),
+		runJobsLimit:    opts.withRunJobsLimit,
+		runJobsInterval: opts.withRunJobInterval,
+	}, nil
+}
+
+// RegisterJob registers a job with the scheduler and persists the job into the repository.
+//
+// • job must be provided and is an implementer of the Job interface.
+//
+// • code must be provided and is used to uniquely distinguish between the same job implementation.
+//
+// The returned JobId is unique to the provide job and code.
+// All options are ignored.
+func (s *Scheduler) RegisterJob(ctx context.Context, j Job, code string, _ ...Option) (JobId, error) {
+	const op = "scheduler.(Scheduler).RegisterJob"
+	if err := validateJob(j); err != nil {
+		return "", errors.Wrap(err, op)
+	}
+	if code == "" {
+		return "", errors.New(errors.InvalidParameter, op, "missing code")
+	}
+
+	jobId, err := job.NewJobId(j.Name(), code)
+	if err != nil {
+		return "", errors.Wrap(err, op)
+	}
+
+	if _, ok := s.registeredJobs.Load(jobId); ok {
+		// Job already registered
+		return JobId(jobId), nil
+	}
+
+	repo, err := s.jobRepoFn()
+	if err != nil {
+		return "", errors.Wrap(err, op)
+	}
+
+	_, err = repo.CreateJob(ctx, j.Name(), code, j.Description())
+	if err != nil && !errors.IsUniqueError(err) {
+		return "", errors.Wrap(err, op)
+	}
+	s.registeredJobs.Store(JobId(jobId), j)
+
+	return JobId(jobId), nil
+}
+
+// UpdateJobNextRun sets the next scheduled run time for the provided jobId
+// to the current database time incremented by the nextRunIn parameter.  If
+// nextRunIn == 0 the job will be available to run immediately.
+//
+// • jobId must be provided and is the id of the job returned by RegisterJob
+//
+// All options are ignored.
+func (s *Scheduler) UpdateJobNextRun(ctx context.Context, jobId JobId, nextRunIn time.Duration, _ ...Option) error {
+	const op = "scheduler.(Scheduler).UpdateJobNextRun"
+	if jobId == "" {
+		return errors.New(errors.InvalidParameter, op, "missing job id")
+	}
+	_, ok := s.registeredJobs.Load(jobId)
+	if !ok {
+		return errors.New(errors.InvalidParameter, op, fmt.Sprintf("job %q not registered", jobId))
+	}
+	repo, err := s.jobRepoFn()
+	if err != nil {
+		return errors.Wrap(err, op)
+	}
+
+	_, err = repo.UpdateJobNextRun(ctx, string(jobId), nextRunIn)
+	if err != nil {
+		return errors.Wrap(err, op)
+	}
+	return nil
+}
+
+// Start begins the scheduling loop that will query the repository for jobs to run and
+// run them in a goroutine.
+func (s *Scheduler) Start(ctx context.Context) {
+	if !s.started.CAS(s.started.Load(), true) {
+		s.logger.Debug("scheduler already started, skipping")
+		return
+	}
+
+	s.baseContext, s.baseCancel = context.WithCancel(ctx)
```

Comment:
Why not use the controller's base context? It seems like this cancel function is only ever called when Shutdown is called; if this is only called when the controller is itself shutdown, you'd already get a canceled base context, so this just becomes superfluous.

Reply:
My thought was to be able to spin the scheduler up and down without needing to cancel the controllers base context.  Not a strict requirement just didn't necessarily want to tie them so strictly together.

Reply:
We can discuss today -- my gut says that the scheduler likely should always be spun up when the controller is active, and it's individual jobs that may or may not need to be spun up.